openvdb_houdini/openvdb_houdini/SOP_OpenVDB_Extrapolate.cc

```diff
@@ -0,0 +1,768 @@
+                        static_cast<double>(evalFloatInst("bgVec3f#", &i, 2, parms.mTime)));
+                    process<FSGridT, openvdb::Vec3DGrid>(parms, lsPrim, fsIsoValue, nullptr /*=maskPrim*/, *extPrim, extBg);
+                    break;
+                } // TYPE_VEC3D
+                case TYPE_VEC3I:
+                {
+                    openvdb::Vec3i extBg(
+                        static_cast<openvdb::Int32>(evalIntInst("bgVec3i#", &i, 0, parms.mTime)),
+                        static_cast<openvdb::Int32>(evalIntInst("bgVec3i#", &i, 1, parms.mTime)),
+                        static_cast<openvdb::Int32>(evalIntInst("bgVec3i#", &i, 2, parms.mTime)));
+                    process<FSGridT, openvdb::Vec3IGrid>(parms, lsPrim, fsIsoValue, nullptr /*=maskPrim*/, *extPrim, extBg);
+                    break;
+                } // TYPE_VEC3I
+            } // eType switch
+        } // end for
+    } else {
+        process<FSGridT>(parms, lsPrim, fsIsoValue, maskPrim);
+    }
+    return true;
+}
+
+
+////////////////////////////////////////
+
+
+//template<typename FSGridT, typename ExtValueT>
+template<typename FSGridT, typename ExtGridT = FSGridT>
+bool
+SOP_OpenVDB_Extrapolate::Cache::process(
+    const FastSweepingParms& parms,
+    hvdb::GU_PrimVDB* lsPrim,
+    typename FSGridT::ValueType fsIsoValue,
+    const hvdb::GU_PrimVDB* maskPrim,
+    hvdb::GU_PrimVDB* exPrim,
+    const typename ExtGridT::ValueType& background)
+{
+    using namespace openvdb::tools;
+
+    using SamplerT = openvdb::tools::GridSampler<ExtGridT, openvdb::tools::BoxSampler>;
+    using ExtValueT = typename ExtGridT::ValueType;
+
+    typename FSGridT::ConstPtr fsGrid = openvdb::gridConstPtrCast<FSGridT>(lsPrim->getConstGridPtr());
+
+    if (parms.mNeedExt) {
+        typename ExtGridT::ConstPtr extGrid = openvdb::gridConstPtrCast<ExtGridT>(exPrim->getConstGridPtr());
+        if (!extGrid) {
+           std::string msg = "Extension grid (" + extGrid->getName() + ") cannot be converted " +
+                             "to the explicit type specified";
+           throw std::runtime_error(msg);
+        }
+        SamplerT sampler(*extGrid);
+        DirichletSamplerOp<ExtGridT> op(extGrid, sampler);
+        using OpT = DirichletSamplerOp<ExtGridT>;
+
+        if (parms.mMode == "fogext" || parms.mMode == "sdfext") {
+            hvdb::Grid::Ptr outGrid;
+            if (parms.mMode == "fogext") {
+                outGrid = fogToExt(*fsGrid, op, background, fsIsoValue, parms.mNSweeps);
+            }
+            else {
+                outGrid = sdfToExt(*fsGrid, op, background, fsIsoValue, parms.mNSweeps);
+            }
+
+            // Create GEO_PrimVDB*
+            std::string primName = parms.mExtPrimName + "Ext";
+            outGrid->setTransform(fsGrid->transform().copy());
+            outGrid->setName(primName);
+            hvdb::createVdbPrimitive(*gdp, outGrid, primName.c_str());
```

Comment:
If I understand properly, we are creatring a new grid called fooExt for the extension.  (And fooSdf down below).  Why not replace the incoming grid with the extension grid?

If I have a Cd grid I want an extended Cd grid out, not an unmodified Cd grid and a CdExt grid?